### DISCUSSION TITLE
Finetuning PaddleOCR

### DISCUSSION THREAD

**Discussion by RRThivyan** [PaddlePaddle/PaddleOCR#13398] [Q&A]
Dear Community,

I need an assistance regarding the finetuing of PaddleOCR. I have successfully fine tuned and created a inference model which contains all 3 pdi files. After this what is the step to do further. Where to save this inference model and how to install this new model?

**Comment by UserWangZz** [comment]
maybe you can reference this doc 
https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_en/detection_en.md#3-evaluation-and-test

**Reply by RRThivyan** [reply]
Hi, the link is for detection. my work in on text recognition. i have followed the (https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_en/recognition_en.md) for recognition, but instead of reading all text, it picks only one letter.

currently i am working on capturing handwritten text from an image. the following is the text captured by default paddleocr model.

[2024/07/16 12:58:17] ppocr DEBUG: dt_boxes num : 22, elapsed : 0.5964698791503906
[2024/07/16 12:58:18] ppocr DEBUG: cls num : 22, elapsed : 0.1217045783996582
[2024/07/16 12:58:19] ppocr DEBUG: rec_res num : 22, elapsed : 1.6230945587158203
Cequua
Address: Wet Rimto
lmakatieity
Age9
M
Date:12-03
Sex:
Rx
(xomrpl
amoxirillin tomg
fe
dan
Seen
loilaay
Physician's Sig
Lic.No
123457
1234567
PTRNo
S2 No

I followed the same instructions mentioned in the link provided but its giving only one letter prediction instead of all text present in the image. the inference_model folder contains all 3 pdi files.

!python3 /content/drive/MyDrive/Verbalyze/IAM_Handwritten_dataset/PaddleOCR/tools/infer/predict_rec.py --image_dir='/content/img1.jpg' --rec_model_dir='/content/drive/MyDrive/Verbalyze/IAM_Handwritten_dataset/PaddleOCR/pretrain_models/inference_model/' --rec_char_dict_path="/content/drive/MyDrive/Verbalyze/IAM_Handwritten_dataset/PaddleOCR/ppocr/utils/en_dict.txt"

[2024/07/16 12:51:53] ppocr WARNING: The first GPU is used for inference by default, GPU ID: 0
[2024/07/16 12:51:55] ppocr INFO: In PP-OCRv3, rec_image_shape parameter defaults to '3, 48, 320', if you are using recognition model with PP-OCRv2 or an older version, please set --rec_image_shape='3,32,320
[2024/07/16 12:51:57] ppocr INFO: Predicts of /content/img1.jpg:('T', 0.9447541236877441)

![image](https://github.com/user-attachments/assets/9c95aea6-f65f-457d-ab3f-bb329f8206f1))

**Reply by UserWangZz** [reply]
Your question is that all handwritten text cannot be recognized using the inference model provided by paddleocr?

**Reply by RRThivyan** [reply]
yes. as you can see from above post, i shared the result of text extraction from base model and inference model. in base model it extracts all text but inference model is fetching only one letter.

**Reply by UserWangZz** [reply]
Can you provide more detailed information? 
Because from the first example you provided, you used a text detection model, a text direction classification model, and a text recognition model.
However, in the second example, you only used the text recognition model to perform text recognition on img1.jpg. Please make sure that the img1 image you provide contains only one instance of text and that the background information is small enough. If not, please use the text detection - text classification - text recognition pipeline to identify img1.

**Reply by RRThivyan** [reply]
oh, so text recognition model only performs for a single text instead of whole image? 
![img1](https://github.com/user-attachments/assets/4ca7c4a6-0395-4eed-9560-ffd8fad69feb)

this is the image i want to extract the text. the base model is not picking some details or predicting wrong text. so i did a fine-tuning.
If I have to extract all text from the image, what is the procedure? any documentation to follow? the above mentioned pipeline is new to me.

**Reply by UserWangZz** [reply]
yes, you need crop text region, then use text recognition model.

**Reply by RRThivyan** [reply]
would you mind sharing the link to the documentation for it? I am new to the pipeline and have to research it before proceeding.

**Reply by UserWangZz** [reply]
> maybe you can reference this doc https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_en/detection_en.md#3-evaluation-and-test

this link is about how to train your text detection model, you need have a labeled dataset about your task. then finetune a text detection model or use offical text detection model if it can satisfy your task. you can use predict_sys.py which is in the tools/infer/ path to connect your pipeline. 
most important make text detection model and recognition model fit your task image.

**Reply by RRThivyan** [reply]
python3 /home/jupyter/PaddleOCR/PaddleOCR/tools/export_model.py -c /home/jupyter/output/v3_en_mobile/config.yml  -o Global.pretrained_model=/home/jupyter/output/v3_en_mobile/best_model/model.pdparams Global.save_inference_dir=gs://paddleocr_ft/inference

i used the above command to infer the model and obtain those 3 pdi files. i also used IAM words dataset to train the model. after inferring, i used predict_rec.py to predict the text from the image as above but as you can see I am getting only a single letter. what is the use of predict_sys.py file? is it to infer like the above command?

**Comment by UserWangZz** [comment]
![image](https://github.com/user-attachments/assets/03a1ff21-8c71-4946-90cf-218166cdba50)
You need to give the recognition model a picture like this

**Reply by RRThivyan** [reply]
But, won't the paddleocr model make the detection of each line and predicts? general model does so. But the predict_rec.py doesn't? only for single line? if we need to pick each line, what is the procedure for that? if I give a single line, it predicts ok, but if I give entire image, no result.
![image](https://github.com/user-attachments/assets/f918b5da-bd20-4abe-a265-33a8b8582f00)

**Reply by UserWangZz** [reply]
We have written a complete pipeline for recognizing the entire image. You need to use the predict_system.py script to specify the specific text detection model, text classification model (if necessary), and text recognition model. In this way, PaddleOCR will first detect instances of the text area, crop the text area from the original input image, and send it to the text recognition model for recognition. This is also the process of general OCR tasks. You need to understand. So for your question, you also need to consider whether the text detection model performs well enough for handwritten text detection.

**Reply by RRThivyan** [reply]
I got it. so before we go to text recognition, we need to finetune text detection model and then only to use predict_system.py. from my observation, it's mentioned to include the inference for detection parameter also. can i use the same dataset for text detection? any particular format for dataset to follow for detection model? I used IAM dataset and it has the boundary box details present. kindly let me know the format of the dataset to be fed into the detection finetuning.
![image](https://github.com/user-attachments/assets/15baa55a-dc03-423a-b1b9-c8a2ce9385f7)

**Reply by UserWangZz** [reply]
https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_en/detection_en.md
please reference doc about text Detection

**Reply by RRThivyan** [reply]
I am trying to use the inference model in the following manner, but the kernel is getting crashed. Though for standard pretrained inference model, its working fine.

![image](https://github.com/user-attachments/assets/436d0819-dcf3-4a51-baeb-290487fc1fa0)
